## Preprocessing

In [1]:

# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
from pathlib import Path


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
application_df = pd.read_csv(Path("/content/drive/MyDrive/Deep-Learning-Challenge/Instructions/charity_data.csv"))
application_df.head()



EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [4]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(["EIN", "NAME"], axis='columns')
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [5]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [6]:
# Look at APPLICATION_TYPE value counts for binning
application_df['APPLICATION_TYPE'].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [7]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_counts= application_df['APPLICATION_TYPE'].value_counts()
application_types_to_replace = list(application_counts[application_counts < 500].index)

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [8]:
# Look at CLASSIFICATION value counts for binning
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [9]:
# You may find it helpful to look at CLASSIFICATION value counts >1
#application_df['CLASSIFICATION'].value_counts() >1
application_df['CLASSIFICATION'].value_counts().loc[lambda x : x>1] 

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [10]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classification_counts= application_df['CLASSIFICATION'].value_counts()
classifications_to_replace = list(classification_counts[classification_counts< 1000].index)
# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [11]:
# Convert categorical data to numeric with `pd.get_dummies`
clean_app_df = pd.get_dummies(application_df)
clean_app_df

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0           1      5000              1                       0   
1           1    108590              1                       0   
2           1      5000              0                       0   
3           1      6692              1                       0   
4           1    142590              1                       0   
...       ...       ...            ...                     ...   
34294       1      5000              0                       0   
34295       1      5000              0                       0   
34296       1      5000              0                       0   
34297       1      5000              1                       0   
34298       1  36500179              0                       0   

       APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                         1                     0                    0   
1                         0                     0                    1   
2                         0                     0                    0   
3                         0                     0                    1   
4                         0                     0                    1   
...                     ...                   ...                  ...   
34294                     0                     0                    0   
34295                     0                     0                    0   
34296                     0                     0                    1   
34297                     0                     0                    0   
34298                     0                     0                    1   

       APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                        0                    0                    0  ...   
1                        0                    0                    0  ...   
2                        0                    1                    0  ...   
3                        0                    0                    0  ...   
4                        0                    0                    0  ...   
...                    ...                  ...                  ...  ...   
34294                    1                    0                    0  ...   
34295                    1                    0                    0  ...   
34296                    0                    0                    0  ...   
34297                    0                    1                    0  ...   
34298                    0                    0                    0  ...   

       INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                      0                       0                         0   
1                      1                       0                         0   
2                      0                       0                         0   
3                      0                       1                         0   
4                      0                       0                         1   
...                  ...                     ...                       ...   
34294                  0                       0                         0   
34295                  0                       0                         0   
34296                  0                       0                         0   
34297                  0                       0                         0   
34298                  0                       0                         0   

       INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                       0                 0                       0   
1                       0                 0                       0   
2                       0                 0                       0   
3                       0                 0                       0   
4                       0                 0                       0   
...                   ...            

In [12]:
# Split our preprocessed data into our features and target arrays
X = clean_app_df.drop('IS_SUCCESSFUL',1)
y = clean_app_df['IS_SUCCESSFUL'].values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  


In [14]:
# Create a MinMaxScaler instances
scaler = MinMaxScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)
X_train_scaled

array([[1.00000000e+00, 2.00100343e-04, 0.00000000e+00, ...,
        0.00000000e+00, 1.00000000e+00, 0.00000000e+00],
       [1.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,
        0.00000000e+00, 1.00000000e+00, 0.00000000e+00],
       [1.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,
        0.00000000e+00, 1.00000000e+00, 0.00000000e+00],
       ...,
       [1.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,
        0.00000000e+00, 1.00000000e+00, 0.00000000e+00],
       [1.00000000e+00, 0.00000000e+00, 0.00000000e+00, ...,
        0.00000000e+00, 1.00000000e+00, 0.00000000e+00],
       [1.00000000e+00, 1.41536185e-05, 0.00000000e+00, ...,
        0.00000000e+00, 1.00000000e+00, 0.00000000e+00]])

## Compile, Train and Evaluate the Model

Optimization Attempts

In [15]:
#Attempt 1 - add another hidden layer

# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
nn1 = tf.keras.models.Sequential()
input_dimentions = len(X_train_scaled[0])

# First hidden layer
nn1.add(tf.keras.layers.Dense(units=10, activation="relu", input_dim=input_dimentions))

# Second hidden layer
nn1.add(tf.keras.layers.Dense(units=6, activation="relu"))

#Third hidden layer
nn1.add(tf.keras.layers.Dense(units=4, activation="relu"))

# Output layer
nn1.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn1.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 10)                440       
                                                                 
 dense_1 (Dense)             (None, 6)                 66        
                                                                 
 dense_2 (Dense)             (None, 4)                 28        
                                                                 
 dense_3 (Dense)             (None, 1)                 5         
                                                                 
Total params: 539
Trainable params: 539
Non-trainable params: 0
_________________________________________________________________


In [16]:
# Compile the model
nn1.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [17]:
2# Train the model
fit_model1 = nn1.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6141 - accuracy: 0.6831
Epoch 2/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5636 - accuracy: 0.7268
Epoch 3/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5567 - accuracy: 0.7317
Epoch 4/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5550 - accuracy: 0.7294
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5532 - accuracy: 0.7311
Epoch 6/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5522 - accuracy: 0.7305
Epoch 7/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5515 - accuracy: 0.7307
Epoch 8/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5508 - accuracy: 0.7313
Epoch 9/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5502 - accuracy: 0.7313
Epoch 10/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5497 - accura

In [18]:
model_loss1, model_accuracy1 = nn1.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss1}, Accuracy: {model_accuracy1}")

268/268 - 0s - loss: 0.5571 - accuracy: 0.7275 - 422ms/epoch - 2ms/step
Loss: 0.5570558905601501, Accuracy: 0.7274635434150696


Adding an additional layer resulted in slightly increased accuracy

In [19]:
#Attempt 2 - use a different activation function + a third layer

# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
nn2 = tf.keras.models.Sequential()
input_dimentions = len(X_train_scaled[0])

# First hidden layer
nn2.add(tf.keras.layers.Dense(units=10, activation="tanh", input_dim=input_dimentions))

# Second hidden layer
nn2.add(tf.keras.layers.Dense(units=6, activation="tanh"))

#Third hidden layer
nn2.add(tf.keras.layers.Dense(units=4, activation="tanh"))

# Output layer
nn2.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn2.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_4 (Dense)             (None, 10)                440       
                                                                 
 dense_5 (Dense)             (None, 6)                 66        
                                                                 
 dense_6 (Dense)             (None, 4)                 28        
                                                                 
 dense_7 (Dense)             (None, 1)                 5         
                                                                 
Total params: 539
Trainable params: 539
Non-trainable params: 0
_________________________________________________________________


In [20]:
# Compile the model
nn2.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [21]:
2# Train the model
fit_model2 = nn2.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5919 - accuracy: 0.7082
Epoch 2/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5679 - accuracy: 0.7257
Epoch 3/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5626 - accuracy: 0.7273
Epoch 4/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5595 - accuracy: 0.7274
Epoch 5/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5575 - accuracy: 0.7276
Epoch 6/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5566 - accuracy: 0.7288
Epoch 7/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5556 - accuracy: 0.7305
Epoch 8/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5550 - accuracy: 0.7303
Epoch 9/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5542 - accuracy: 0.7299
Epoch 10/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5537 - accura

In [22]:
model_loss2, model_accuracy2 = nn2.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss2}, Accuracy: {model_accuracy2}")

268/268 - 1s - loss: 0.5535 - accuracy: 0.7307 - 534ms/epoch - 2ms/step
Loss: 0.5535073280334473, Accuracy: 0.7307288646697998


changing the activation function increased the accuracy of the model a little more

In [23]:
#Attempt 3 - increasing units and epochs

# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
nn3 = tf.keras.models.Sequential()
input_dimentions = len(X_train_scaled[0])

# First hidden layer
nn3.add(tf.keras.layers.Dense(units=15, activation="tanh", input_dim=input_dimentions))

# Second hidden layer
nn3.add(tf.keras.layers.Dense(units=10, activation="tanh"))

#Third hidden layer
nn3.add(tf.keras.layers.Dense(units=6, activation="tanh"))

# Output layer
nn3.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn3.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_8 (Dense)             (None, 15)                660       
                                                                 
 dense_9 (Dense)             (None, 10)                160       
                                                                 
 dense_10 (Dense)            (None, 6)                 66        
                                                                 
 dense_11 (Dense)            (None, 1)                 7         
                                                                 
Total params: 893
Trainable params: 893
Non-trainable params: 0
_________________________________________________________________


In [24]:
# Compile the model
nn3.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [25]:
2# Train the model
fit_model3 = nn3.fit(X_train_scaled, y_train, epochs=200)

Epoch 1/200
804/804 [==============================] - 2s 2ms/step - loss: 0.5974 - accuracy: 0.6967
Epoch 2/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5698 - accuracy: 0.7224
Epoch 3/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5634 - accuracy: 0.7247
Epoch 4/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5611 - accuracy: 0.7265
Epoch 5/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5583 - accuracy: 0.7275
Epoch 6/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5568 - accuracy: 0.7276
Epoch 7/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5552 - accuracy: 0.7284
Epoch 8/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5538 - accuracy: 0.7292
Epoch 9/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5528 - accuracy: 0.7282
Epoch 10/200
804/804 [==============================] - 1s 2ms/step - loss: 0.5526 - accura

In [26]:
model_loss3, model_accuracy3 = nn3.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss3}, Accuracy: {model_accuracy3}")

268/268 - 0s - loss: 0.5559 - accuracy: 0.7299 - 384ms/epoch - 1ms/step
Loss: 0.555949866771698, Accuracy: 0.729912519454956


Accuracy was slightly reduced compared to the 2nd attempt

In [27]:
#Attempt 4 - deacreasing units and epochs, trying a different activation function
nn4 = tf.keras.models.Sequential()
input_dimentions = len(X_train_scaled[0])

# First hidden layer
nn4.add(tf.keras.layers.Dense(units=10, activation="sigmoid", input_dim=input_dimentions))

# Second hidden layer
nn4.add(tf.keras.layers.Dense(units=8, activation="sigmoid"))

# Output layer
nn4.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn4.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_12 (Dense)            (None, 10)                440       
                                                                 
 dense_13 (Dense)            (None, 8)                 88        
                                                                 
 dense_14 (Dense)            (None, 1)                 9         
                                                                 
Total params: 537
Trainable params: 537
Non-trainable params: 0
_________________________________________________________________


In [28]:
# Compile the model
nn4.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [29]:
2# Train the model
fit_model4 = nn4.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 2s 2ms/step - loss: 0.6508 - accuracy: 0.6347
Epoch 2/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5916 - accuracy: 0.7163
Epoch 3/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5794 - accuracy: 0.7227
Epoch 4/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5759 - accuracy: 0.7231
Epoch 5/50
804/804 [==============================] - 1s 1ms/step - loss: 0.5738 - accuracy: 0.7233
Epoch 6/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5724 - accuracy: 0.7246
Epoch 7/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5707 - accuracy: 0.7250
Epoch 8/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5693 - accuracy: 0.7247
Epoch 9/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5677 - accuracy: 0.7257
Epoch 10/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5664 - accuracy: 0.7262

In [30]:
model_loss4, model_accuracy4 = nn4.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss4}, Accuracy: {model_accuracy4}")

268/268 - 0s - loss: 0.5573 - accuracy: 0.7254 - 368ms/epoch - 1ms/step
Loss: 0.5572835803031921, Accuracy: 0.7253644466400146


This decreased accuracy to the original model

In [31]:
# Export our model to HDF5 file
nn2.save('/content/drive/MyDrive/Deep-Learning-Challenge/Instructions/AlphabetSoupCharity_Optimized.h5')

Overview of the analysis: 

Explain the purpose of this analysis.

-We want to predict whether or not applicants for funding will be successful.


Results: Using bulleted lists and images to support your answers, address the following questions.



Data Preprocessing

What variable(s) are considered the target(s) for your model?

-'IS_SUCCESSFUL'

What variable(s) are considered to be the features for your model?

-'APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION', 'USE_CASE', 'ORGANIZATION', 'STATUS', 'INCOME_AMT', 'SPECIAL_CONSIDERATIONS', 'ASK_AMT'

What variable(s) are neither targets nor features, and should be removed from the input data?

-'EIN', 'NAME'


Compiling, Training, and Evaluating the Model

How many neurons, layers, and activation functions did you select for your neural network model, and why?

-2 layers, with 10 and 6 nuerons, and the relu activation function. 
-I used the tendor flow playground to see what would lead to a smaller loss


Were you able to achieve the target model performance?

-No, but I was pretty close with 73% accuracy

What steps did you take to try and increase model performance?

- Instead of using standard scaler, I tried the minmax scaler
- I made 4 attempts to increase the accuracy
1. I added an additional hidden layer
2. I tried using a different activation function
3. I added more units and epochs
4. I reduced units, epochs, and tried another activation function





Summary: Summarize the overall results of the deep learning model. Include a recommendation for how a different model could solve this classification problem, and explain your recommendation.

I think the most effective way to find the best model would be to utilize the auto optimization feature. I feel that the Relu function is likely the best to use as its generally the ideal function. Potantially further normalizing the data would result in better accuracy as well. 

